android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt

```diff
@@ -518,83 +542,122 @@ class RNGestureHandlerModule(reactContext: ReactApplicationContext?)
     handlerFactories.firstOrNull { it.type == handler.javaClass } as HandlerFactory<T>?
 
   private fun <T : GestureHandler<T>> onHandlerUpdate(handler: T, motionEvent: MotionEvent) {
+    // onUpdate
+
     if (handler.tag < 0) {
       // root containers use negative tags, we don't need to dispatch events for them to the JS
       return
     }
     if (handler.state == GestureHandler.STATE_ACTIVE) {
       val handlerFactory = findFactoryForHandler(handler)
 
-      if (handler.usesDeviceEvents) {
+      if (handler.actionType == GestureHandler.ACTION_TYPE_REANIMATED_WORKLET) {
```

Comment:
All those changes are needed due to different handling of events in Fabric? If so, maybe it would be good to give a quick explanation why is it like that now for other libraries maintainers, or is it something that will happen only in this library?

Reply:
**1) Reanimated**

In the old architecture, Reanimated overwrites RCTEventDispatcher with REAEventDispatcher to intercept all direct events (see [here](https://github.com/software-mansion/react-native-reanimated/blob/a870aa28092322b627695f8cf2ea0dce4db34a53/ios/REAEventDispatcher.m#L8-L12)). This allows Gesture Handler to integrate with Reanimated without defining a hard dependency, simply by sending regular direct events.

In the new architecture, the event system was mainly ported to C++ so we can no longer overwrite RCTEventDispatcher. Instead, we need to dispatch events directly to Reanimated internals.

One more thing to note here is that in Reanimated for Paper there is no feedback if Reanimated did or did not handle specific event, so effectively they are handled twice (once by Reanimated and once by React Native). We plan to address this issue simultaneously with the adoption of Fabric.

**2) Animated.event with useNativeDriver: true**

Currently, `NativeAnimated[Turbo]Module` is a `RCTEventDispatcherObserver` (see [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/React/CoreModules/RCTEventDispatcher.mm#L117-L119)). Basically, it listens to all events going through `RCTEventDispatcher` but cannot prevent them from being sent to JS side. That's why in such case, similarly as for Reanimated, we would like to invoke NativeAnimated[Turbo]Module directly instead of flooding the app with unnecessary events.

**3) Direct vs. device events**

On Paper, the old Gesture Handler API (e.g. `PanGestureHandler`) uses direct events in order to invoke JS callbacks from the native side. When designing API for Gesture Handler 2.0 it turned out to be a limitation, so the new API (i.e. `GestureDetector`) uses device events instead.

With Fabric, at some point we decided to use only device events, both for old and new API. That's because I've already managed to get device events working in Gesture Handler but also it was kinda unclear how to send direct events from Obj-C. It looks like for this purpose we need to use C++ EventEmitter that we can access from the provided ShadowNode. However, currently Gesture Handler is set up as a NativeModule (in Java/Obj-C) so we would need to migrate it first to C++ and use JSI in order to pass ShadowNodes from JS.

Effectively, using direct or device events should be no difference for the library user as this seems to be an internal implementation detail of our library. Nevertheless, we will research into ways of sending direct events like we used to do on Paper.

Reply:
Maybe it would be nice to add this e.g. to the description of PR so it does not get lost?

Reply:
> However, currently Gesture Handler is set up as a NativeModule (in Java/Obj-C) so we would need to migrate it first to C++ and use JSI in order to pass ShadowNodes from JS.

> Why is Gesture Handler NativeModule (in Java/Obj-C) was not changed to TurboModule?

Reply:
Hey @rashmi-priya! The reason why we didn't migrate RNGH NativeModule to a TurboModule is the fact that NativeModules still work on the new architecture. The goal of the migration was to add Fabric support to RNGH 2.x, so the package itself supports all versions of RN 0.63.0+ as well as the old and the new architecture.

TurboModules are backward-compatible (meaning they work as NativeModules on the old architecture) but we couldn't just rewrite NativeModule into a TurboModule because it would not compile on older versions of RN (e.g. 0.64).

Another option was to conditionally enable NativeModule on Paper and TurboModule on Fabric. However, in this variant we would duplicate code and the secondary goal for this PR was to introduce as few changes as possible.

We decided to leave NativeModule as it is and address this issue in the future. Maybe now it's a good time to do so. 😄 

I hope this answers your questions.

Reply:
Thanks for the clarification.